### PR TITLE
feat(cudf): Add support for `DATE_ADD` and `DATE_TRUNC`

### PR DESCRIPTION
**Pull request by mattgara**
* Add support for `DATE_ADD` and `DATE_TRUNC`, for `DATE` types

* Add test coverage

--------
Follow-up on #16634, dependent on that merging first.

### CODE REVIEW

**Comment by karthikeyann** [comment]
Hi @kjmph, if you have bandwidth, could you take a look at this PR? The changed files overlap with areas you have edited recently, so your context would be helpful. Thanks.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+  switch (unit_) {
+    case DateTimeUnit::kSecond:
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::SECOND, stream, mr);
+    case DateTimeUnit::kMinute:
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::MINUTE, stream, mr);
+    case DateTimeUnit::kHour:
+      return cudf::datetime::floor_datetimes(
```
Comment (kjmph) [inline velox/experimental/cudf/expression/DateTruncFunction.cpp:155]: Can we gate or implement timezone-aware timestamp truncation here?

Velox `date_trunc(timestamp, ...)` uses `truncateTimestamp(timestamp, unit, getTimeZoneFromConfig(config))`, so when `adjust_timestamp_to_timezone` is enabled it truncates in the session timezone and converts back to GMT. This cuDF path truncates the raw cuDF timestamp and has no access to `QueryConfig`.

For example, with `session_timezone = Asia/Kolkata` and `adjust_timestamp_to_timezone = true`, `2026-01-02 00:45:00 UTC` is local `2026-01-02 06:15:00`. `date_trunc('hour', ts)` should truncate to local `06:00`, which is `2026-01-02 00:30:00 UTC`. Raw cuDF/UTC truncation returns `2026-01-02 00:00:00 UTC`.

This affects `hour` and larger units. `day/week/month/quarter/year` differ for most non-UTC zones, and `hour` differs for zones with non-hour offsets or around some DST boundaries.

I think the fix is either:
- plumb `QueryConfig`/timezone into `DateTruncFunction` and implement Velox’s timezone-adjusted semantics, or
- reject timestamp `date_trunc` when `adjust_timestamp_to_timezone` is enabled so this falls back to CPU.

Suggested test if this path implements timezone-aware truncation:

```cpp
TEST_F(CudfFilterProjectTest, dateTruncTimestampUsesSessionTimezone) {
  auto data = makeRowVector(
      {"event_ts"},
      {makeFlatVector<Timestamp>(
          {Timestamp(1767314700, 0)}, // 2026-01-02 00:45:00 UTC
          TIMESTAMP())});

  auto plan = PlanBuilder()
                  .values({data})
                  .project({
                      "date_trunc('hour', event_ts) AS h",
                      "date_trunc('day', event_ts) AS d",
                  })
                  .planNode();

  auto expected = makeRowVector(
      {"h", "d"},
      {
          // Asia/Kolkata local 06:15 -> 06:00 == 2026-01-02 00:30:00 UTC.
          makeFlatVector<Timestamp>({Timestamp(1767313800, 0)}, TIMESTAMP()),
          // Asia/Kolkata local day start == 2026-01-01 18:30:00 UTC.
          makeFlatVector<Timestamp>({Timestamp(1767292200, 0)}, TIMESTAMP()),
      });

  AssertQueryBuilder(plan)
      .config(core::QueryConfig::kSessionTimezone, "Asia/Kolkata")
      .config(core::QueryConfig::kAdjustTimestampToTimezone, "true")
      .assertResults(expected);
}
```

**Review by kjmph — CHANGES_REQUESTED** [review]

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 1 above)
Comment (shrshi) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp:155]: This is an excellent point. AFAIK, cuDF does not have timezone-aware timestamp truncation/rounding APIs at the C++ level. It, however, has the building blocks - we can build a timezone transition table, convert the UTC timestamps to local timestamps, and then truncate. The implementation feels non-trivial though, and I would support deferring the implementation to a later PR. 
For now, +1 for the other suggestion from Kyle - reject timestamp date_trunc when adjust_timestamp_to_timezone is enabled so this falls back to CPU.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            intType,
+            stream,
+            mr);
+        monthsToSubtract = cudf::binary_operation(
+            monthIndex->view(),
+            quarterStart->view(),
+            cudf::binary_operator::SUB,
+            intType,
+            stream,
+            mr);
```
Comment (shrshi) [inline velox/experimental/cudf/expression/DateTruncFunction.cpp:235]: IIUC, L204-225 are doing `monthsToSubtract = monthIndex - (monthIndex / 3) * 3`. Can we replace these three kernels with MOD `monthsToSubtract = monthIndex % 3`?

File: velox/experimental/cudf/expression/DateTruncFunction.cpp [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::SECOND, stream, mr);
+    case DateTimeUnit::kMinute:
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::MINUTE, stream, mr);
+    case DateTimeUnit::kHour:
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::HOUR, stream, mr);
+    case DateTimeUnit::kDay: {
+      auto dayCol = castToDay(inputCol);
```
Comment (shrshi) [inline velox/experimental/cudf/expression/DateTruncFunction.cpp]: Do we need `castToDay` here? Since this input is already in `TIMESTAMP_DAYS`, can we directly return the input column?

File: velox/experimental/cudf/expression/DateTruncFunction.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  explicit DateTruncFunction(const std::shared_ptr<velox::exec::Expr>& expr);
+
+  ColumnOrView eval(
+      std::vector<ColumnOrView>& inputColumns,
+      rmm::cuda_stream_view stream,
+      rmm::device_async_resource_ref mr) const override;
+
+ private:
+  functions::DateTimeUnit unit_;
+  bool isTimestamp_{false};
```
Comment (shrshi) [inline velox/experimental/cudf/expression/DateTruncFunction.h]: I don't think we are using isTimestamp_ and isDate_ anywhere. Can we remove these variables?

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 1 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp:155]: Thanks, well spotted. I've addressed this by taking the fallback path (implementing timezone-aware cuDF truncation in this PR would be large scope creep.) The cuDF FilterProject selection now rejects `date_trunc(timestamp, ...)` when `QueryConfig::kAdjustTimestampToTimezone` is enabled, so those expressions stay on CPU and preserve session-timezone semantics. I also added test coverage for this.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 2 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp:235]: Good point. Replaced the `FLOOR_DIV`/`MUL`/`SUB` sequence with `monthIndex % 3`, so quarter truncation computes `monthsToSubtract` with one `MOD` operation.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 3 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp]: Agreed. For the `day` unit, the cuDF path now returns the input view directly when it is already `TIMESTAMP_DAYS`.

File: velox/experimental/cudf/expression/DateTruncFunction.h (hunk 4 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.h]: Yes, I think these must of snuck in during the splitting of the two PRs and were never cleanup, thanks!

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 1 above)
Comment (shrshi) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp:155]: @mattgara Can you please open an issue for this so that we don't lose track of this discussion?

**Review by shrshi — APPROVED** [review]

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 1 above)
Comment (devavret) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp:155]: A relevant PR to track is https://github.com/facebookincubator/velox/pull/17899 . It implements timezone support and also passes along query context. @mattgara you can wait until that is merged to implement this here.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+          inputCol, cudf::datetime::rounding_frequency::SECOND, stream, mr);
+    case DateTimeUnit::kMinute:
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::MINUTE, stream, mr);
+    case DateTimeUnit::kHour:
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::HOUR, stream, mr);
+    case DateTimeUnit::kDay:
+      if (inputCol.type() == dayType) {
+        return inputCol;
```
Comment (devavret) [inline velox/experimental/cudf/expression/DateTruncFunction.cpp]: I don't think you're allowed to return a view from anything other than input columns (as-in input to the expression). Here the column can be non-input like it could be computed by another date function which is part of the subexpression. to be safe, materalize and return.

File: velox/experimental/cudf/exec/CudfFilterProject.cpp [hunk 6]
```diff
@@ -120,7 +148,19 @@ bool canBeEvaluatedByCudf(
...
       std::move(exprsCopy), &precompileCtx, lazyDereference);
 
+  const core::QueryConfig defaultQueryConfig = core::QueryConfig({});
+  const core::QueryConfig& queryConfig =
+      queryCtx ? queryCtx->queryConfig() : defaultQueryConfig;
+  const bool adjustTimestampToTimezone =
+      queryConfig.adjustTimestampToTimezone();
+
   for (const auto& e : exprSet->exprs()) {
+    if (adjustTimestampToTimezone && containsTimezoneSensitiveDateTrunc(e)) {
```
Comment (devavret) [inline velox/experimental/cudf/exec/CudfFilterProject.cpp:155]: I see a problem here now that may not be solvable in this PR. We rely on cudf expressions in CudfHiveConnector as well, in the remaining filter. But there we never do a canBeEvaluatedByCudf check. That should be a Presto fix.

**Review by devavret — CHANGES_REQUESTED** [review]
mostly lgtm. there's a couple of places which can blow up in certain cases. One of this is a fix i'm requesting in this PR and for the other, we need an issue for.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 1 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp:155]: @devavret I think this is required for TPC-DS benchmarking, so probably worth landing this sooner.

Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp:155]: @shrshi  Please see issue added as comment here: https://github.com/facebookincubator/velox/issues/17410#issuecomment-4919379077

File: velox/experimental/cudf/exec/CudfFilterProject.cpp (hunk 6 above)
Comment (mattgara) [reply velox/experimental/cudf/exec/CudfFilterProject.cpp:155]: Good catch, I've added this as a comment in this existing issue: https://github.com/facebookincubator/velox/issues/17410#issuecomment-4919379077

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 5 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp]: Good point. I updated this path to materialize instead of returning the `column_view` directly.

This pattern is likely to appear in other parts and is worth auditing separately. For example, similar pattern in [`LogicalFunction`](https://github.com/facebookincubator/velox/blob/235d3df9d5999ce093db236ef54a2bc1d6dd9564/velox/experimental/cudf/expression/ExpressionEvaluator.cpp#L825-L829) and [`CoalesceFunction`](https://github.com/facebookincubator/velox/blob/235d3df9d5999ce093db236ef54a2bc1d6dd9564/velox/experimental/cudf/expression/ExpressionEvaluator.cpp#L1239-L1246). 

Those two locations are using patterns such as

```
  auto input = asView(inputColumns[0]);
...
  return ColumnOrView(input);
```
  or:

```
  ColumnOrView result = asView(inputColumns[0]);
...
  return result;
```

I'm not sure but these may also need to be materialized to be safe. (I'm not proposing doing this in this PR, however, just suggesting we do an audit for it.)

File: velox/experimental/cudf/expression/DateTruncFunction.cpp [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::MINUTE, stream, mr);
+    case DateTimeUnit::kHour:
+      return cudf::datetime::floor_datetimes(
+          inputCol, cudf::datetime::rounding_frequency::HOUR, stream, mr);
+    case DateTimeUnit::kDay:
+      if (inputCol.type() == dayType) {
+        return std::make_unique<cudf::column>(inputCol, stream, mr);
+      }
+      return castDaysToOutput(castToDay(inputCol));
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/DateTruncFunction.cpp]: For timestamp inputs this casts to `TIMESTAMP_DAYS` (and the week/month/quarter/year paths do the same via `castToDay`). libcudf timestamp casts truncate toward zero rather than floor, so a pre-epoch (negative) timestamp truncates to the wrong day and can cross a month/year boundary — e.g. `date_trunc('month', 1969-12-31 20:00:00)` would give `1970-01-01` instead of `1969-12-01`. The sub-day units already use `floor_datetimes` and are correct.

`makeTimestampExtractVectors` is entirely post-1970, so this isn't exercised. Could you floor timestamps to day (`floor_datetimes(inputCol, rounding_frequency::DAY)`) before the calendar math, and add a pre-1970 timestamp to the differential test to confirm?

File: velox/experimental/cudf/expression/DateTruncFunction.cpp [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    VELOX_CHECK(
+        isTimestamp, "date_trunc {} requires timestamp input", *unitString);
+  }
+}
+
+ColumnOrView DateTruncFunction::eval(
+    std::vector<ColumnOrView>& inputColumns,
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr) const {
+  auto inputCol = asView(inputColumns[0]);
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/DateTruncFunction.cpp:119]: `canEvaluate` doesn't reject a constant data input (`inputs()[1]`), but this does `asView(inputColumns[0])` unconditionally — a constant data input yields an empty `inputColumns` and indexes out of bounds. Constant folding removes the all-constant case in practice, but `DateAddFunction::canEvaluate` guards the analogous case explicitly (returns false when both value and date are constants). Worth the same defensive check here for symmetry, or a `VELOX_CHECK` on `inputColumns` size.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+  };
+  auto castDaysToOutput =
+      [&](std::unique_ptr<cudf::column> daysCol) -> ColumnOrView {
+    if (daysCol->type() == outputType) {
+      return daysCol;
+    }
+    return cudf::cast(daysCol->view(), outputType, stream, mr);
+  };
+
+  auto makeScalar = [&](int32_t value) {
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/DateTruncFunction.cpp]: `oneScalar`/`threeScalar`/`negOneScalar` are constant int32 scalars re-allocated on every `eval` (per batch). `DatePlusIntervalFunction` caches its constant scalars as members in the constructor to avoid per-batch reallocation — same idea would apply here.

File: velox/experimental/cudf/expression/DateTruncFunction.h [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+  explicit DateTruncFunction(const std::shared_ptr<velox::exec::Expr>& expr);
+
+  ColumnOrView eval(
+      std::vector<ColumnOrView>& inputColumns,
+      rmm::cuda_stream_view stream,
+      rmm::device_async_resource_ref mr) const override;
+
+ private:
+  functions::DateTimeUnit unit_;
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/DateTruncFunction.h]: nitpick: `unit_` is missing uniform initialization here (`functions::DateTimeUnit unit_{};`), while `DateAddFunction.h` uses `unit_{}`.

File: velox/experimental/cudf/expression/ExpressionEvaluator.h [hunk 11]
```diff
@@ -61,6 +62,15 @@ inline std::vector<cudf::column_view> tableViewToColumnViews(
   return result;
 }
 
+// Throws a VeloxUserError with userMessage if any non-null entry of cond is
+// false. cond must be a BOOL8 column. Does nothing for empty or all-null
+// columns.
+void checkAllTrue(
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/ExpressionEvaluator.h:68]: `checkAllTrue` is now a shared helper declared here but defined in `DatePlusIntervalFunction.cpp`. Should the definition move to `ExpressionEvaluator.cpp` with the other shared helpers? Also, its doc comment is duplicated between this declaration and the definition — keep it in the header only.

File: velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+      mr);
+
+  auto inRange = cudf::binary_operation(
+      geMin->view(),
+      leMax->view(),
+      cudf::binary_operator::LOGICAL_AND,
+      boolType,
+      stream,
+      mr);
+  checkAllTrue(inRange->view(), "date_add value is out of range", stream, mr);
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp]: Out-of-range errors differ by path: the column path throws a user error here ("date_add value is out of range"), while the literal path throws `VELOX_UNSUPPORTED` ("Value should be in range ...") via `checkValueInInt32Range`. Worth aligning the category and message.

**Review by karthikeyann — COMMENTED** [review]
**GPU `date_add` / `date_trunc` — approve with comments.**

Overall this looks good: the timezone fallback, `kDay` materialization, quarter `% 3`, and the earlier unused-member cleanup are all in, `mr`/`stream` are threaded through every cuDF call, and the int32 overflow handling on both the literal and column paths is careful.

Main items left:
- Pre-epoch timestamps: the timestamp -> day cast truncates toward zero, so day/week/month/quarter/year truncation is likely wrong for negative timestamps (untested — test data is all post-1970). See inline.
- Defensive symmetry with `DateAdd` for a constant data input.
- Minor: `checkAllTrue` definition location, per-batch scalar allocation, `unit_` init, and out-of-range error consistency.

Nothing blocking.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 7 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp]: Good catch. I changed the timestamp calendar-unit path to floor timestamps to day precision with `cudf::datetime::floor_datetimes(..., rounding_frequency::DAY)` before converting to `TIMESTAMP_DAYS` for the day/week/month/quarter/year math. I also added a pre-1970 timestamp/date row to the shared Presto/generic differential `date_trunc` coverage, plus direct SparkSQL timestamp assertions for the same pre-epoch day/month/quarter/year cases.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 8 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp:119]: Done. `DateTruncFunction::canEvaluate` now rejects a constant data input so the cuDF evaluator is not selected for an all-constant expression, and `eval` has a defensive `inputColumns` size check before reading `inputColumns[0]`.

File: velox/experimental/cudf/expression/DateTruncFunction.cpp (hunk 9 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.cpp]: Done. `oneScalar_`, `threeScalar_`, and `negOneScalar_` are now cached on `DateTruncFunction`.

File: velox/experimental/cudf/expression/DateTruncFunction.h (hunk 10 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/DateTruncFunction.h]: Fixed. `unit_` now uses uniform initialization.

File: velox/experimental/cudf/expression/ExpressionEvaluator.h (hunk 11 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/ExpressionEvaluator.h:68]: Done. I moved `checkAllTrue` into `ExpressionEvaluator.cpp` with the other shared evaluator helpers and removed the duplicate implementation comment from `DatePlusIntervalFunction.cpp`; the documentation now remains on the header declaration.

File: velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp (hunk 12 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp]: Done. The literal `date_add` path now uses the same Velox user error and message as the column path (`date_add value is out of range`), and the existing literal out-of-range test assertion was updated.

File: velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+    cudf::duration_scalar<cudf::duration_D> days(
+        checkedScaleValue(literalValue_, scale),
+        literalValueIsValid_,
+        stream,
+        mr);
+    return cudf::binary_operation(
+        dateCol, days, cudf::binary_operator::ADD, outType, stream, mr);
+  }
+
+  auto daysInt = scaleToInt32(*valueCol, scale, stream, mr);
```
Comment (devavret) [inline velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp]: nit: you can throw when the day is out of range but do you need to do that for rows where the date itself is null? I think this is similar to the issue we've faced earlier about handling subexpressions that can throw but they'd never be evaluated if the outer expression doesn't need to calculate it for a certain row. something to keep in mind, not requesting for this PR.

**Review by devavret — APPROVED** [review]

**Comment by devavret** [comment]
@mattgara can you please resolve conflicts

**Comment by mattgara** [comment]
> @mattgara can you please resolve conflicts

Should be done now.

File: velox/experimental/cudf/exec/CudfFilterProject.cpp [hunk 14]
```diff
@@ -50,6 +53,31 @@ void debugPrintTree(
   }
 }
 
+bool isTimezoneSensitiveDateTrunc(
+    const std::shared_ptr<velox::exec::Expr>& expr) {
+  if (!DateTruncFunction::canEvaluate(expr) ||
```
Comment (karthikeyann) [inline velox/experimental/cudf/exec/CudfFilterProject.cpp]: `second` and `minute` truncation are timezone-independent in Velox, but this sends them to CPU whenever timezone adjustment is enabled. Could this inspect the constant unit and fall back only for `hour` and larger units?

File: velox/experimental/cudf/exec/CudfFilterProject.cpp [hunk 15]
```diff
@@ -50,6 +53,31 @@ void debugPrintTree(
...
+      !expr->inputs()[1]->type()->isTimestamp()) {
+    return false;
+  }
+
+  auto cudfFunction = createCudfFunction(expr->name(), expr);
```
Comment (karthikeyann) [inline velox/experimental/cudf/exec/CudfFilterProject.cpp]: Could this avoid calling `createCudfFunction` during eligibility? Constructing `DateTruncFunction` performs three device allocations and a stream synchronization, so a query destined for CPU fallback can still stall or fail on GPU allocation. Would side-effect-free registry metadata work here?

File: velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+
+// Multiplies value by scale and returns the result as int32_t, throwing if
+// either value or value*scale overflow int32_t.
+void checkDateAddValueInInt32Range(int64_t value) {
+  VELOX_USER_CHECK(
+      value == static_cast<int32_t>(value), "date_add value is out of range");
+}
+
+int32_t checkedScaleValue(int64_t value, int32_t scale) {
+  checkDateAddValueInInt32Range(value);
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp:71]: Could the CPU and GPU overflow policies be aligned? This path rejects `value * scale` overflow, while the CPU path checks only `value` before scaling, so enabling cuDF changes behavior for the same query. A boundary parity test would make the intended behavior explicit.

**Review by karthikeyann — APPROVED** [review]

File: velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp (hunk 13 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp]: I've addressed the local case where `checkValueRange` now treats a row with a null date as valid for range-check purposes, so an out-of-range value in that row does not throw.

The broader case where an outer expression does not select an otherwise non-null row still requires "active-row" propagation (or some way to obtain this information) and remains outside this PR.

File: velox/experimental/cudf/exec/CudfFilterProject.cpp (hunk 14 above)
Comment (mattgara) [reply velox/experimental/cudf/exec/CudfFilterProject.cpp]: Implemented, gating now checks for the units.

File: velox/experimental/cudf/exec/CudfFilterProject.cpp (hunk 15 above)
Comment (mattgara) [reply velox/experimental/cudf/exec/CudfFilterProject.cpp]: Removed this `createCudfFunction` call in favour of matching the function name/handle.

File: velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp (hunk 16 above)
Comment (mattgara) [reply velox/experimental/cudf/expression/prestosql/DateAddFunction.cpp:71]: Aligned the GPU path to now match the CPU policy. It checks that the raw `BIGINT` fits in `int32_t`, then applies unit scaling after conversion instead of rejecting scaled overflow. Also added a test covering this boundary.
